### DISCUSSION TITLE
Multiplication remarks

### DISCUSSION THREAD

**Discussion by Samuel-DEVULDER** [spotlessmind1975/ugbasic#268] [Optimizations]
I had a look at the multiplication code generation:
https://github.com/spotlessmind1975/ugbasic/blob/919069944b9175531e9c79f5f6abe2bc6303561e/ugbc/src/targets/common/_infrastructure.c#L1630-L1642

For me it look like the 32bit multiplication is "buggy" (well there is a warning in that case). It calls `cpu_math_mul_16bit_to_32bit` hence omitting half the data of the 32bit source. Then $12345678 * $00000001 will result in $00001234 or $00004567 or $00000000 depending on the words effectively lost in the process. This won't be close to the expected result. I think there must be some kind of 32x32 -> 64 signed mul, or simply a 32x32->32 unsigned mul (see below). 

Another remarks appears when multiplying  16x16 and 8x8. I noticed that we allocate a 32 or 16 bit result, call a signed multiplication if input is signed, and drop the upper half of the result. This is following the mathematical rules. Great!

But did you know that the lower half of unsigned multiplications matches the lower half of the signed ones ? Try it out: $FF * $FF = $FE01 whose lower half ($01) matches the signed result (-$01 * -$01), the same appear with $FF * $12 = $11EE (-$12 = $EE), or $FFFF * $FFFF =$FFFE0001, or $FFFF * $1234 = $1233EDCC (-$1234 = $EDC), and actually any pair of signed/unsigned numbers.  This is a nice hacker trick !

This mean we can simply call unsigned multiplication in that case which will be much faster (easy optim)

And do go further,  we don't even need to call a fully unsigned multiplication as only the lower half is needed. So only 8x8->8, 16x16->16 and 32x32->32 unsigned muls are actually needed for this routine. This will speedup implementations with shift+add as only 8 or 16 bits numbers hare to be handled without the need for extra bits.

**Comment by spotlessmind1975** [comment]
Hi @Samuel-DEVULDER , **thank you for the detailed analysis!**

I think it's "buggy" too, but that's a known fact. In fact, as you too have guessed, to multiply two 32-bit integers you need a 64-bit integer as the destination of the multiplication result, and when I wrote the generator I had no routines capable of working with 64 bits, and there is no I still have them. The notice would serve precisely to report the fact.

> Then $12345678 * $00000001 will result in $00001234 or $00004567 or $00000000 depending on the words effectively lost in the process. 

In general, the omission concerns the upper part of the operation and not the lower part. So the right example would be:

$12345678 * $00000001 = $00005678 * $00000001 = $00005678

It is wrong anyway, I agree with you, but at least it look like an "overflow" (truncation to 16 bits).

> Another remarks appears when multiplying 16x16 and 8x8. I noticed that we allocate a 32 or 16 bit result, call a signed multiplication if input is signed, and drop the upper half of the result. 

It doesn't seem to me that we throw away the upper part of the result, unless you intend that, outside the multiplication, there is a casting towards a less capacious type. Am I missing something?

>  This mean we can simply call unsigned multiplication in that case which will be much faster (easy optim)

The rule of the low half seems to me to have exploited it in the multiplication routines written directly in assembly (the various `cpu_math_div_16bit_to_16bit.asm` and so on): in practice the `*_SIGNED` version performs, in advance, the two's complement for negative numbers before proceed to the unsigned multiplication, and then reapply the product of the signs at the end of the multiplication. I don't know if you meant this, but I hope so.

> And do go further, we don't even need to call a fully unsigned multiplication as only the lower half is needed. So only 8x8->8, 16x16->16 and 32x32->32 unsigned muls are actually needed for this routine. 

Um ... but this multiplication 42 (8bit) x 127 (8 bit) = 5334 (16 bit), if we applied the "truncation", it would give a wrong result: 214 (8 bit). Is there something wrong?

ps. I have activated the [discussion](https://github.com/spotlessmind1975/ugbasic/discussions) section because I think that this type of analisys should have more visibility, and the possibility to share the analysis with other users. Only if you agree, we could move the issue there. As soon as we arrive to one or more real "activities", we could open one or more issue. What do you think?

**Comment by Samuel-DEVULDER** [comment]
> It doesn't seem to me that we throw away the upper part of the result, unless you intend that, outside the multiplication, there is a casting towards a less capacious type. Am I missing something?

Yeah, when the result is assigned back to a variable of the same type as the input the upper part of the result is lost (truncated), and in that case it would have been sufficient, and faster, to just perform multiplication in an unsigned way as the lower part of multiplication is the same if we consider both input as unsigned.

> Um ... but this multiplication 42 (8bit) x 127 (8 bit) = 5334 (16 bit), if we applied the "truncation", it would give a wrong result: 14 (8 bit). Is there something wrong?

If the result need 16 bit yes you are right truncation is not good (c is wrong). But it is the result one get after the assignment of the expression back to a byte value. 

```
VAR a AS SIGNED BYTE
VAR b AS SIGNED BYTE
VAR c AS SIGNED BYTE

a = 42
b = 127
c = a*b : REM c is ko because 16 bits are required to store the result (5334)
```
But if the result doesn't require 16 bits, for instance -42 * 3 = -126 = $FF82 which is truncated back to $82 (-126),
```
VAR a AS SIGNED BYTE
VAR b AS SIGNED BYTE
VAR c AS SIGNED BYTE

a = -42
b = 3
c = a*b : REM c is okay with 8 bits (-126)
```
then we can perform a*b as an unsigned multiplication (no need to make -42 become 42 and apply the sign back after): (-42) * 3 = (214) * 3 = 642 = $0282 which is truncated back to $82. 

Do you see ? Same result. No need to convert the arguments to positive value, perform an unsigned MUL, and apply the sign back to the result. Doing the multiplication in a unsigned way without converting the value to positive produces the very  same truncated result.

This was the point I wanted to spot. This is actually a reason why there is no 8x8->8 signed MUL on the 6809 as it is just the same as the truncation of the 8x8->16 unsigned MUL already present.

May be this optimisation is not possible yet in the compiler as it require to know in advance that the result will be truncated by the assignment. 

In this approach the type of operations to performs on the RHS is "driven" by the (already fixed) size of LHS. If LHS is n-bits wide, then everything, including multiplications, can be truncated/limited to n-bits without affecting the result. This is classic modular arithmetic. And if the case of truncation to "half size", the multiplication can be performed unsigned because modular arithmetic doesn't really have sign (-1 and 255 are both the same in 8 bit world).

I hope I am clear. (Not sure )

> ps. I have activated the discussion section because I think that this type of analisys should have more visibility, and the possibility to share the analysis with other users. Only if you agree, we could move the issue there. As soon as we arrive to one or more real "activities", we could open one or more issue. What do you think?

Yes, why not. I don't know this feature , but a dedicated place to discussions is way better than along with "issues" as these aren't real issues anyway :)

**Comment by spotlessmind1975** [comment]
> Yeah, when the result is assigned back to a variable of the same type as the input the upper part of the result is lost (truncated), and in that case it would have been sufficient, and faster, to just perform multiplication in an unsigned way as the lower part of multiplication is the same if we consider both input as unsigned.

You are clearly right, even if I don't quite understand why it should happen.

> But it is the result one get after the assignment of the expression back to a byte value.

Ok, it is more clear now! In my humble opinion, the program has a bug.
The correct version of the program is:

    VAR a AS SIGNED BYTE 
    VAR b AS SIGNED BYTE
    VAR c AS SIGNED WORD

    a = 42
    b = 127
    c = a*b : REM now c is ok!

An even more correct version would have to omit the typing of the variable `c`, so that ugBASIC initializes it with the correct size (16 bits). A side note: this should be the correct approach because we are talking about a BASIC-like language. The fact that you can type implicitly is actually a necessity to optimize but the language shouldn't dictate that.

> Do you see ? Same result. No need to convert the arguments to positive value, perform an unsigned MUL, and apply the sign back to the result. Doing the multiplication in a unsigned way without converting the value to positive produces the very same truncated result.

**Wow, this is really cool!** Could you hypothesize a refactoring of a signed / unsigned 8x8 bit multiplication, in order to make me understand how it could be done? That would certainly be a nice optimization because, in fact, I could do without the 2's complement of each operand (which takes up time and space).

> This was the point I wanted to spot. This is actually a reason why there is no 8x8->8 signed MUL on the 6809 as it is just the same as the truncation of the 8x8->16 unsigned MUL already present. May be this optimisation is not possible yet in the compiler as it require to know in advance that the result will be truncated by the assignment.
> In this approach the type of operations to performs on the RHS is "driven" by the (already fixed) size of LHS. If LHS is n-bits wide, then everything, including multiplications, can be truncated/limited to n-bits without affecting the result. This is classic modular arithmetic. And if the case of truncation to "half size", the multiplication can be performed unsigned because modular arithmetic doesn't really have sign (-1 and 255 are both the same in 8 bit world).

So let's see if I understand correctly. A signed NxN -> N multiplication is equivalent to an unsigned NxN-> 2N multiplication with N truncation. Obviously, you guessed correctly: this being an LR parser, it has a hard time "anticipating" typing. In fact, I don't even think it's possible, in the sense that every expression that contains a multiplication NxN eventually gets a 2N expression, which is then assigned (truncated). So I'm afraid there is no way to get help from the compiler.

On the other hand, we have the infixed notation of multiplication (`MUL`) which might help us.

> Yes, why not. 

I did it! 👍

**Comment by Samuel-DEVULDER** [comment]
> On the other hand, we have the infixed notation of multiplication (MUL) which might help us.

Yeah, this is the way to introduce this speedup. 👍

MUL x,y can perform a sizeof(x)-unsigned-multiplication over y and always get the correct result regardless of x or y being signed. This means: 
* if x is 8 bit, a 8x8->8 unsigned-mul, 
* if x is 16 bits, a 16x16->16 unsigned-mul, and finally 
* if x is 32 a 32x32->32 unsigned-mul. 

Of couse this will require some cpu_math_mul_Nbit_to_Nbit(), which can be highly optimized in asm (it is just a single ASM instruction for the 6809 for 8bit), but can be emulated via cpu_math_mul_Nbit_to_2Nbit() + truncation as a start.

**Reply by spotlessmind1975** [reply]
Hi @Samuel-DEVULDER, **it is a very good suggestion**. I opened an issue on that.

**Comment by spotlessmind1975** [comment]
Hi @Samuel-DEVULDER , in the meanwhile I fixed the multiplication between different signed bitwidth variables, at this commit:
2475e48
